### PR TITLE
Canbus + MCP2515

### PR DESCRIPTION
**Pull request by danielschramm**
## Description:
This one replaces https://github.com/esphome/esphome-docs/pull/648
Some schematics and an example were added.
Both schematics are tested by me and work well.

**mvturnho also merged the changes so #648 and this one are now exact duplicates.**

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#1384

## Checklist:

  - [x] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [x] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

**Comment by pergolafabio** [comment]
Hi, @danielschramm 

Do you also have a schematic for the nodemcu v3 ? I want to try out the PR
And is there also a debug option to see all can messages for receiving? Because I don't know yet what my messages are...


Thnx in advance

**Comment by danielschramm** [comment]
> Hi, @danielschramm
> 
> Do you also have a schematic for the nodemcu v3 ? I want to try out the PR
> And is there also a debug option to see all can messages for receiving? Because I don't know yet what my messages are...
> 
> Thnx in advance

Hi @pergolafabio ,

Nodemcu v3 and Wemos D1 should work the same.
Schematic is visible in the preview:
https://deploy-preview-823--esphome.netlify.app/_images/canbus_mcp2515_resistor.png

Just checked all signal pins. They are the same for Wemos and Nodemcu v3
Signal : GPIO : Wemos : Nodemcu v3
SCK : GPIO 16 : D0 : D0
MOSI : GPIO5 : D1 : D1
MISO : GPIO4 : D2 : D2
CS : GPIO 14 : D5 : D5

Depending on the log level the details of the received frames are shown in the Esphome log.
Higher level also enables the Frame content.

In the Base Config section of the docs is also an example how to display the received data for only one specific ID.
https://deploy-preview-823--esphome.netlify.app/components/canbus.html

Setting up 2 EspHome CAN nodes will give you full testing capabilities without any special CAN-Debugging hardware.

Please let me know if everything is working well.

Daniel

**Comment by pergolafabio** [comment]
hi @danielschramm  , ok, clear, gonna test out asap

so for logging/candump, i just need this :   ?

```
# Enable logging
logger:
 level: DEBUG
```

**Comment by danielschramm** [comment]
@pergolafabio 
level: VERY_VERBOSE
Should enable the output of received payload.

**Comment by pergolafabio** [comment]
perfect, thnx for feedback, appriciated

**Comment by pergolafabio** [comment]
Hi @danielschramm , last question, I have esphome running as addon, but how can I install your custom version? Components I can run as custom, but how todo it with ad addon?


Edit: nm, found it

**Comment by pergolafabio** [comment]
> > Hi, @danielschramm
> > Do you also have a schematic for the nodemcu v3 ? I want to try out the PR
> > And is there also a debug option to see all can messages for receiving? Because I don't know yet what my messages are...
> > Thnx in advance
> 
> Hi @pergolafabio ,
> 
> Nodemcu v3 and Wemos D1 should work the same.
> Schematic is visible in the preview:
> https://deploy-preview-823--esphome.netlify.app/_images/canbus_mcp2515_resistor.png
> 
> Just checked all signal pins. They are the same for Wemos and Nodemcu v3
> Signal : GPIO : Wemos : Nodemcu v3
> SCK : GPIO 16 : D0 : D0
> MOSI : GPIO5 : D1 : D1
> MISO : GPIO4 : D2 : D2
> CS : GPIO 14 : D5 : D5
> 
> Depending on the log level the details of the received frames are shown in the Esphome log.
> Higher level also enables the Frame content.
> 
> In the Base Config section of the docs is also an example how to display the received data for only one specific ID.
> https://deploy-preview-823--esphome.netlify.app/components/canbus.html
> 
> Setting up 2 EspHome CAN nodes will give you full testing capabilities without any special CAN-Debugging hardware.
> 
> Please let me know if everything is working well.
> 
> Daniel

@danielschramm 

i am wiring up the mcp and nodecmuv3
seems the nodecmu3 doesnt have an 5v output, only 3,3
will that work too? if yes, do i still need that 4k7 resistor then?

**Comment by danielschramm** [comment]
> @danielschramm
> 
> i am wiring up the mcp and nodecmuv3
> seems the nodecmu3 doesnt have an 5v output, only 3,3
> will that work too? if yes, do i still need that 4k7 resistor then?

Hi @pergolafabio here are a lot of hardware discussions in #648.
It will most likely not work with 3.3V.
But the nodemcu3 has 5V in Pin "VIN":
https://shopofthings.ch/wp-content/uploads/2017/11/nodemcuv3_pinout.png

**Comment by pergolafabio** [comment]
> > @danielschramm
> > i am wiring up the mcp and nodecmuv3
> > seems the nodecmu3 doesnt have an 5v output, only 3,3
> > will that work too? if yes, do i still need that 4k7 resistor then?
> 
> Hi @pergolafabio here are a lot of hardware discussions in #648.
> It will most likely not work with 3.3V.
> But the nodemcu3 has 5V in Pin "VIN":
> https://shopofthings.ch/wp-content/uploads/2017/11/nodemcuv3_pinout.png

i thought that was an input , not an output

**Comment by danielschramm** [comment]
> > But the nodemcu3 has 5V in Pin "VIN":
> 
> i thought that was an input , not an output

Same as on D1 Mini directly connected to 5V on the USB so can be used for powering the module if no USB connected or to power other components with 5V from USB,

**Comment by pergolafabio** [comment]
@danielschramm , i did the wiring, i have no canbus H/L connected yet, just put the code below in loopback mode to test the MCP with very_verbose log ...
installed this version : esphome_version: 'danielschramm:canbus'

dot this output looks ok ?
also, that log file, is that saved somewhere? easier to debug for later

code: 

```
spi:
  id: McpSpi
  clk_pin: GPIO16
  mosi_pin: GPIO5
  miso_pin: GPIO4
  
canbus:
  - platform: mcp2515
    id: my_mcp2515
    spi_id: McpSpi
    cs_pin: GPIO14
    can_id: 1
    mode: loopback  
```
 
logger:
  level: VERY_VERBOSE

```
[15:21:48][VV][spi:016]: Disabling SPI Chip on pin 14...
[15:21:48][VV][spi:095]: Enabling SPI Chip on pin 14...
[15:21:48][VV][spi:090]:     TX 0b10100000 (0xA0)
[15:21:48][VV][spi:093]:     RX 0b00000000 (0x00)
[15:21:48][VV][spi:090]:     TX 0b00000000 (0x00)
[15:21:48][VV][spi:093]:     RX 0b00000000 (0x00)
[15:21:48][VV][spi:016]: Disabling SPI Chip on pin 14...
[15:21:48][VV][spi:095]: Enabling SPI Chip on pin 14...
[15:21:48][VV][spi:090]:     TX 0b10100000 (0xA0)
[15:21:48][VV][spi:093]:     RX 0b00000000 (0x00)
[15:21:48][VV][spi:090]:     TX 0b00000000 (0x00)
[15:21:48][VV][spi:093]:     RX 0b00000000 (0x00)
[15:21:48][VV][spi:016]: Disabling SPI Chip on pin 14...
[15:21:48][VV][spi:095]: Enabling SPI Chip on pin 14...
[15:21:48][VV][spi:090]:     TX 0b10100000 (0xA0)
[15:21:48][VV][spi:093]:     RX 0b00000000 (0x00)
[15:21:48][VV][spi:090]:     TX 0b00000000 (0x00)
[15:21:48][VV][spi:093]:     RX 0b00000000 (0x00)
[15:21:48][VV][spi:016]: Disabling SPI Chip on pin 14...
```

**Comment by danielschramm** [comment]
Setting every component to very_verbose gives much to many information.
In the short log there are only SPI Logs included which are not very interesting for MCP2515 and Canbus.
https://esphome.io/components/logger.html
You can set only the interesting elements to higher log level and reduce all other to a minimum.

**Comment by pergolafabio** [comment]
Ok, will have a look at it...
But how to test my board? Without a canbus attached to H/L ? What's that loopback all about?

**Comment by danielschramm** [comment]
> Ok, will have a look at it...
> But how to test my board? Without a canbus attached to H/L ? What's that loopback all about?

I did not check loopback functionality right now.
I have a running CAN-Bus and USB-Can reference hardware and did the testing against that.
For real and not too expensive testing I would recommend setting up 2 ESP8266 with MCP2515.
Can Termination on both ends can be set via Jumper and only direct connection is needed.

Loopback mode is for special integrity testing of the CAN Interface itself and I really see no application for Loopback in combination with Esphome since in Loopback no data is sent over the lines.

**Comment by pergolafabio** [comment]
Ok, just gonna hook up the 2 cables to my canbus, normally nothing can go wrong ;-)

Thnx for feedback.. any idea when this PR will be merged? Or can I run your dev mode?

**Comment by danielschramm** [comment]
> Ok, just gonna hook up the 2 cables to my canbus, normally nothing can go wrong ;-)

> Thnx for feedback.. any idea when this PR will be merged? Or can I run your dev mode?

Currently the maintainers seem pretty busy since a lot of good Pulls are waiting for review.
I think every feedback on the stability of the new Code will help.
So I'm really looking forward to get feedback from you.
Also if there are any bugs in the documentation.

**Comment by pergolafabio** [comment]
> > Ok, just gonna hook up the 2 cables to my canbus, normally nothing can go wrong ;-)
> 
> > Thnx for feedback.. any idea when this PR will be merged? Or can I run your dev mode?
> 
> Currently the maintainers seem pretty busy since a lot of good Pulls are waiting for review.
> I think every feedback on the stability of the new Code will help.
> So I'm really looking forward to get feedback from you.
> Also if there are any bugs in the documentation.

ok, great, i have hooked it up, but not sure how to get the canbus dump, just wanna watch the canbus feed
how can i debug that specific component?

i have now listenonly mode

```
spi:
  id: McpSpi
  clk_pin: GPIO16
  mosi_pin: GPIO5
  miso_pin: GPIO4
  
canbus:
  - platform: mcp2515
    id: my_mcp2515
    spi_id: McpSpi
    cs_pin: GPIO14
    can_id: 1
    mode: LISTENONLY  
 
logger:
  level: VERY_VERBOSE
  logs:
    esphome.components.canbus: VERY_VERBOSE
    esphome.components.mcp2515: VERY_VERBOSE
```

**Comment by pergolafabio** [comment]
or maybe i have a wire issue, i have used that VIN pin, but if i put a multimeter on it, i only count like 1 V output on that vcc pin

**Comment by pergolafabio** [comment]
ok, succes!! seems i use the wrong output pin, that VIN pin you told earlier is an input pin, i needed to use the VU PIN
now i see messages like below! so it works

any idea how i can use that info?  in my case , the message below is when i turned on a light
the light as had id = 41:00 , thats in message data[1] and data[2]

all my lights are like 41:00 or 41:05 or 42:07 .....
data[3] is the on/off, so if i turn **off** a light wit id 41:00,   => message.data[3]=00  , if i turn on light with id 41:00  => message.data[3]=01

how can i use that info to HA? maybe create a binary sensor? that use that binary sensor in esp as a value template for my lights?

thnx for the info!!

```
[22:35:24][D][canbus:046]: received can message can_id=-2130641152 size=4
[22:35:24][V][canbus:052]:   can_message.data[0]=00
[22:35:24][V][canbus:052]:   can_message.data[1]=41
[22:35:24][V][canbus:052]:   can_message.data[2]=00
[22:35:24][V][canbus:052]:   can_message.data[3]=01

[22:35:49][D][canbus:046]: received can message can_id=-2130641152 size=4
[22:35:49][V][canbus:052]:   can_message.data[0]=00
[22:35:49][V][canbus:052]:   can_message.data[1]=41
[22:35:49][V][canbus:052]:   can_message.data[2]=00
[22:35:49][V][canbus:052]:   can_message.data[3]=00
```

**Comment by pergolafabio** [comment]
@danielschramm , see my previous post about the binary sensor
and also a new question, how can i use that array of data to call a service
in HA, i have template lights, switches, see below example, those are the same values i see on canbus

i think i need to use the on_frame service, like described here : https://deploy-preview-823--esphome.netlify.app/components/canbus.html#on-frame

but how can i use that data/array?

for me its only import to read from the canbus, with the canbus information i want todo stuff in HA
maybe in a later phase, i actually gonna write to the canbus

so questions:

1) the binary sensor option in esphome , turn on/off based on canbus array (see previous post)
2) call a service in HA based on the canbus array

thnx a lot !!

```
#example in HA to turn on a light on relay module 43 , id 10 , 01 => power on
        - service: shell_command.dobiss
          data_template:
            module: 43
            id: 10
            power: 01
```

**Comment by danielschramm** [comment]
@pergolafabio 
Please disable "LISTENONLY". That disables the generation of ACK signals.
Listenonly is only useful if you just want to passively sniff a BUS. For Example you intend to hack your vehicles CAN bus.
Otherwise if you only have 2 nodes the transmitting node will soon get a tx buffer overflow and stop transmitting.

To get Services run in HA it's best to map the the received data to a template sensor, like in the Binary Sensor example.
With this sensor you can trigger an automation in HA.

**Comment by pergolafabio** [comment]
Yeah, my first purpose is indeed to hack into the canbus

But I am not sure how to make that binary sensors, I have an array of date, not sure how I can use that info to make a binary sensor in esp 
I need that data[3] , that's my bit that turns something on/off


Can you post an example ? So I can start with testing your component?

Thnw

**Comment by danielschramm** [comment]
> Yeah, my first purpose is indeed to hack into the canbus
> 
> But I am not sure how to make that binary sensors, I have an array of date, not sure how I can use that info to make a binary sensor in esp
> I need that data[3] , that's my bit that turns something on/off
> 
> Can you post an example ? So I can start with testing your component?
> 
> Thnw

In thie Example on CAN id 0x100  Byte 3 -> 0 = off,  1 =on
The template Sensor is just defined and it's state ist updated within the lambda on message received.

binary_sensor:
  - platform: template
    name: "CAN Bus Button"
    id: "can_bus_button"

canbus:
  - platform: mcp2515
    id: my_mcp2515
    spi_id: McpSpi
    cs_pin: GPIO14
    can_id: 4
    bit_rate: 125kbps
    on_frame:
    **- can_id: ${0x100}**
      then:
        - lambda: |-
            switch(**x[3]**) {
              case **0x0**: id(can_bus_button).publish_state(false); break; // button release
              case **0x1**: id(can_bus_button).publish_state(true); break;  // button down
            }

**Comment by pergolafabio** [comment]
Great, i try it out later
I am new to esphome, for every yaml change, you need to always compile and upload?

**Comment by danielschramm** [comment]
> Great, i try it out later
> I am new to esphome, for every yaml change, you need to always compile and upload?

yes

**Comment by pergolafabio** [comment]
> > Yeah, my first purpose is indeed to hack into the canbus
> > But I am not sure how to make that binary sensors, I have an array of date, not sure how I can use that info to make a binary sensor in esp
> > I need that data[3] , that's my bit that turns something on/off
> > Can you post an example ? So I can start with testing your component?
> > Thnw
> 
> In thie Example on CAN id 0x100 Byte 3 -> 0 = off, 1 =on
> The template Sensor is just defined and it's state ist updated within the lambda on message received.
> 
```
binary_sensor:

platform: template
name: "CAN Bus Button"
id: "can_bus_button"
canbus:

platform: mcp2515
id: my_mcp2515
spi_id: McpSpi
cs_pin: GPIO14
can_id: 4
bit_rate: 125kbps
on_frame:
- can_id: ${0x100}
then:
- lambda: |-
switch(x[3]) {
case 0x0: id(can_bus_button).publish_state(false); break; // button release
case 0x1: id(can_bus_button).publish_state(true); break; // button down
}

```
ok, i am testing this, not fully understanding your, code, what is that 0x100 in your example?
if you look in my canbus dump example below, those are 4 different lights that are currently off
data[1] together with data[2] , so 4100 or 4101 or 4102 or 4910 ... those are the id's of my lights
data[3] is representing 00 or 01 , on or off

so i am struggling with this : 
```
     - can_id: ${0x100}
      then:
```
what do i need to put there? as you can see , the can_id's below are always the same in my case, it seems it also starts with a minus - sign .. , do i also need to put that - sign ?

[19:31:21][D][canbus:046]: received can message can_id=-2130641152 size=4
[19:31:21][V][canbus:052]:   can_message.data[0]=00
[19:31:21][V][canbus:052]:   can_message.data[1]=41
[19:31:21][V][canbus:052]:   can_message.data[2]=00
[19:31:21][V][canbus:052]:   can_message.data[3]=00


[19:31:21][D][canbus:046]: received can message can_id=-2130641152 size=4
[19:31:21][V][canbus:052]:   can_message.data[0]=00
[19:31:21][V][canbus:052]:   can_message.data[1]=41
[19:31:21][V][canbus:052]:   can_message.data[2]=01
[19:31:21][V][canbus:052]:   can_message.data[3]=00

[19:31:21][D][canbus:046]: received can message can_id=-2130641152 size=4
[19:31:21][V][canbus:052]:   can_message.data[0]=00
[19:31:21][V][canbus:052]:   can_message.data[1]=41
[19:31:21][V][canbus:052]:   can_message.data[2]=02
[19:31:21][V][canbus:052]:   can_message.data[3]=00

**Comment by pergolafabio** [comment]
also if i use:
```
    on_frame:
    - can_id: ${0x100}
```
i get:

```
      Expected integer, but cannot parse ${0x100} as an integer.
      can_id: ${0x100} [source /config/esphome/esp.yaml:44]
```

if i use this:
```
    on_frame:
    - can_id: -2130641152
```
or:
```
    on_frame:
    - can_id: 2130641152
```

i get:


```
      value must be at least 1.
      can_id: -2130641152 [source /config/esphome/esp.yaml:44]
```
or
```
      value must be at most 4096.
      can_id: 2130641152 [source /config/esphome/esp.yaml:44]
```

**Comment by pergolafabio** [comment]
ok, created a pull request, i changed the value of  integers , no more 4096 as max, but also a negative value now with a higher integer
now it seems to work : https://github.com/danielschramm/esphome/pull/1

now i just need to identify the light id itself within those can_ids , i can maybe to that with if/else constructions

**Comment by pergolafabio** [comment]
ok, found it , its working, probably code can be improved :)

can you merge that request with higer id ?
here is my working code:

```
    on_frame:
    - can_id: -2130641152
      then:
        - if:
            condition:
              lambda: 'return x[1] == 0x41 and x[2] == 0x00;'
          
            then:
              - lambda: |-
                  switch(x[3]) {
                    case 0x0: id(can_bus_button).publish_state(false); break; // button release
                    case 0x1: id(can_bus_button).publish_state(true); break;  // button down
                  }   
```

**Comment by danielschramm** [comment]
Currently only Standard IDs are supported by the code.
The used libs support the extended IDs.
I think it should be higher priority to get Canbus support with standard IDs merged and later add an update to support extended IDs.

Or is it absolutely necessary for your application to have extended IDs?

**Comment by pergolafabio** [comment]
> Currently only Standard IDs are supported by the code.
> The used libs support the extended IDs.
> I think it should be higher priority to get Canbus support with standard IDs merged and later add an update to support extended IDs.
> 
> Or is it absolutely necessary for your application to have extended IDs?

yeah, i think so , i have no control over the id's that are being used, thats why i changed the code to make it higher as 4096
now it works

**Comment by timvdsm** [comment]
Hi Daniel, 
Many thanks for all your hard work.
I am also testing this with an ESP8266 (nodemcu v3) + MCP2515. 
In the beginning there was no problem, but now, after adding more code, I receive an error when flashing my esp ... 
As I'm capturing the status of all my lights + covers my yaml file is around 750 code lines.

This is the error I receive:

Using ‘COM3’ as serial port.
Connecting…
Detecting chip type… ESP826 %)Writing at 0x00044000… (100 %)Wrote 421520 bytes (285503 compressed) at 0x00000000 in 7.0 seconds (effective 478.3 kbit/s)…
Hash of data verified.

Leaving…
Hard Resetting…
Done! Flashing is complete!

Showing logs:
[13:07:05]
[13:07:05]SDK:2.2.2-dev(38a443e)/Core:2.7.3-3-g2843a5ac=20703003/lwIP:STABLE-2_1_2_RELEASE/glue:1.2-30-g92add50/BearSSL:5c771be
[13:07:05][I][logger:166]: Log initialized
[13:07:05][C][ota:366]: There have been 1 suspected unsuccessful boot attempts.
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:05]Fatal exception 29(StoreProhibitedCause):
[13:07:13]Fa
[13:07:13] ets Jan 8 2013,rst cause:4, boot mode:(3,6)
[13:07:13]
[13:07:13]wdt reset
[13:07:13]load 0x4010f000, len 3584, room 16
[13:07:13]tail 0
[13:07:13]chksum 0xb0
[13:07:13]csum 0xb0
[13:07:13]v2843a5ac
[13:07:13]~ld
[13:07:13]

Is this related to a storage issue on the ESP? Or is there something else you are aware off?

**Comment by danielschramm** [comment]
> Hi Daniel,

Hi @timvdsm could you please create an issue on the fork that we can discuss details there: 
https://github.com/danielschramm/esphome/issues
That will keep this Pull request thread clean for directly code related discussions.
Could you please include your whole yaml file to that issue that I can directly test it against my hardware.
Thanks, Daniel

File: components/canbus.rst [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+Each canbus platform extends this configuration schema.
+
+.. code-block:: yaml
+
+    # Example configuration entry
+    canbus:
+      - platform: ...
+        can_id: 4
+        use_extended_id: False
+        bit_rate: 50KBPS
```
Comment (glmnet) [inline components/canbus.rst]: keep initial examples as simple as possible, if there are defaults for the values don't include in this first example

File: components/canbus.rst [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+          - lambda: |-
+              std::string b(x.begin(), x.end());
+              ESP_LOGD("can id 500", "%s", &b[0] );
+
+Configuration variables:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
+- **can_id** (**Required**, integer): default *can id* used for transmitting frames.
+- **use_extended_id** (*Optional*, boolean): default *False* identifies the type of *can_id*: 
+  *False*: Standard 11 Bit IDs, *True*: Extended 29Bit ID
```
Comment (glmnet) [inline components/canbus.rst]:
```suggestion
  *False*: Standard 11 bits IDs, *True*: Extended 29 bits ID
```

File: components/canbus.rst [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+This automation will be triggered when a can frame is  received. A variable ``x`` of type
+``std::vector<uint8_t>`` is passed to the automation for use in lambdas.
+
+.. code-block:: yaml
+
+    canbus:
+      - platform: ...
+        on_frame:
+        - can_id: 43 # the remote sender can_id
```
Comment (glmnet) [inline components/canbus.rst]: this comment is a bit misleading, the can_id is the message id, not sender :) 
also, a good remark here will be to explain somewhere what will happen if you send a message to a can_id you listen to, will you catch it too?

File: components/canbus.rst [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+          then:
+            - if:
+                condition:
+                  lambda: 'return x[0] == 0x11;'
```
Comment (glmnet) [inline components/canbus.rst]: this example might want to check on x's length before accessing the index, canbus messages can be zero DLC right?

``` e.g.
lambda: 'return (x.size() > 0) ? x[0] == 0x11 : false;' 
```

File: components/canbus.rst [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+            - light.turn_on:
+                id: light_1
+                brightness: !lambda "return (float) x[0]/255;"
+        - can_id: 502
+            then:
+            - light.turn_on:
+                id: light_1
+                brightness: !lambda "return (float) x[0]/255;"
+
+Wireing options
```
Comment (glmnet) [inline components/canbus.rst]:
```suggestion
Wiring options
```

File: components/canbus.rst [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+
+Standard vs. Extended ID
+------------------------
+| Standard IDs and Extended IDs can coexist on the same segment.
+| It is important to know that for example Standard 0x123 and Extended 0x123 are different addesses.
+| This example shows how the different ID types are used in the configuration for transmission and receiving.
+| For the IDs decimal or hexadecimal notation is possible:
+| 0x000 - 0x7ff / 0-2047 for Standard IDs only.
+| 0x00000000 - 0x1fffffff / 0-536870911 for Extended IDs.
```
Comment (glmnet) [inline components/canbus.rst]: Change format of this, you can use a note call out, please review the contributing docs for formatting

**Review by glmnet — CHANGES_REQUESTED** [review]

File: components/canbus.rst (hunk 6 above)
Comment (danielschramm) [reply components/canbus.rst]: Thanks for the hints, I will check them tomorrow, it's just time to go to bed in Germany ;-)

File: components/canbus.rst (hunk 3 above)
Comment (danielschramm) [reply components/canbus.rst]: Changed the comment:
_the_ received _can_id_

And added this to the description:
_Messeges this node sents to te same ID will not show up as reveived messages._

File: components/canbus.rst (hunk 1 above)
Comment (danielschramm) [reply components/canbus.rst]: removed both default lines

File: components/canbus.rst (hunk 4 above)
Comment (danielschramm) [reply components/canbus.rst]: Good hint.
Changed all other examples as well.

File: components/canbus.rst (hunk 6 above)
Comment (danielschramm) [reply components/canbus.rst]: OK, I kept the Chapter but renamed to "Extended ID" and made part of the Text a note.

"please review the contributing docs for formatting"
Are there other things to be optimized? I'm not exactly sure what you mean. Thanks.

**Review by glmnet — APPROVED** [review]

**Pull request merged**